docs/proposals/optional_application_loging.md

```diff
@@ -0,0 +1,106 @@
+Optional Application Logging
+============================
+
+The EFK aggregated logging stack by default collects application logs with
+fluentd from STDOUT of each container managed by OpenShift. Currently only way
+of opting out is based on a node label and ansible variable
+`openshift_logging_fluentd_nodeselector` as described in the
+[documentation](https://docs.openshift.org/latest/install_config/aggregate_logging.html).
+This proposal discusses some ways of a finer grained level of control over
+which container logs are stored in Elasticsearch.
+
+
+## Fluentd pipeline
+
+The container logs can be obtained from one of these supported logging drivers
+- json-file - logs read from `/var/log/containers/*` mounted in fluentd
+              container
+- journald
+
+Latest development shifted recommended docker driver from journald to json-file,
+therefore, the proposed implementation has focused on usage of json-file as
+well. Rationale being, journald currently doesn't match performance expectations
+and desired throttling control for larger clusters.
+
+Each container on the host has its own json structured log file
+- `/var/log/containers/[pod]_[namespace]_[container]-[docker_id].log`
+
+Every line is treated as a separate log message and processed in the cascade
+of fluentd filters with currently only supported sink Elasticsearch.
+A log line doesn't contain OpenShift metadata, this is added to the message
+in the fluentd `k8_metadata_filter` plugin.
+
+#### Dynamic Control of Optional Logging
+
+By 'dynamic control' we understand the ability of an operator or a user to
+opt-in or opt-out certain OpenShift object from aggregated logging without
+having to redeploy fluentd.
+
+This approach has immediate effect, offers very granular level of control,
+is flexible, and doesn't bring any measurable latency for other containers logs.
+It requires additional resources and brings an overhead to the API server, where
+the fluentd input plugin has to `watch` certain API endpoints.
+
+Following schema describes how this can be implemented as an opt-out kind:
+- a user will label a namespace or a pod with `aggregated_logging_opt_out=true`
+- the fluentd will exclude log files based on this information
+![fluentd optin pipeline](https://raw.githubusercontent.com/wozniakjan/origin-aggregated-logging/proposal/optional_logging/docs/proposals/optin_pipeline.png)
+
+One way how to achieve this could be creating additional layer of symlinks to
+the log files. Currently the log files in `/var/log/containers/` are symlinks
+to files in `/var/log/pods/`, those are symlinks to files in `/var/lib/docker/containers/`,
+and finally those are UTF-8 formatted text files. We could create symlinks to
+`/var/log/containers/*` in a different directory, `/var/log/containers/optin/*`
+for instance. Then a daemon watching the OpenShift APIs and keep the symlinks
```

Comment:
Where does this daemon live and how is it deployed?  @pweil- Is this a case for a custom controller?

Reply:
The way I tried to implement a prototype was simply having a bash script running in the background inside each fluentd container. It did work but our fluentd json-file pipeline currently tags container logs using their full path and depends on this for further processing on multiple places. To hack around this, I used the `rewrite_tag_filter` back to to the old know container log path just to see it works. It does, but also at that point, I thought this is probably too hackerish and decided to try another approach, modifying the tail plugin to see if that leads to something cleaner.

Custom controller could be very elegant way, it didn't occur to me and I haven't explored that path yet. Another viable approach to the solution could be a separate container in fluentd pod mounting mentioned hostPaths, watching the OpenShift API, and keeping the symlinks up to date.

Reply:
As you have discovered, the file path is tied to the fluentd tags which is tied to the workflow.  It will be risky to change that.  Why isn't using `exclude_path` sufficient?  Because it isn't dynamic?

Reply:
I was trying to only document multiple options and compare them. But I personally would prefer a dynamic control, as it appears a bit more elegant for the end user (either the `in_tail` plugin changes or additional layer of symlinks). On the other hand, it will require more maintenance from us in the future.

Reply:
The daemon must be deployed on each node then?

Have you considered instead specifying "destinations" instead of "source"? i.e.
if namespace has label/annotation directing it to "exclude" logs, then logs go to `/dev/null`.
if namespace has label to include logs, then logs go to default service `logging-es.logging.svc.service.cluster.local`.
if namespace is infrastructure ns, then logs go to ops service `logging-ops-es`.
if user wants to send logs to external service, then logs go to this service specified in the annotation.

i.e. all logs are collected, decision whether to forward them somewhere or not and where to forward to is made within the collector.

Reply:
> The daemon must be deployed on each node then?

yes, every mentioned solution affects fluentd only, therefore, must be deployed on each node (the prototype is a part of fluentd container)

> Have you considered instead specifying "destinations" instead of "source"

That originally wasn't the scope of this effort, please read [here](https://docs.google.com/document/d/1m9-LXr-UThs9WhRfkFRbwf1Lppxa5Opw_PYRY1wuDbY/edit#heading=h.s6k4plkzrdji). From my understanding, we wanted a way of dealing with pods that went awry and tried to log excessive amounts of data (pod exclusion) and a feature for openshift.io (namespace exclusion). I was not aware we would like to change the routing logic for `ops` and `non-ops`. But in case we do, I can go back to drawing board and play around with it a bit more. Or maybe it can have its own proposal as a distinct feature.

Reply:
on changing the routing loging for `ops` and non `ops` I was maybe going too far.
The key thing I meant was that if destinations were being considered instead of source then 1 daemon per cluster might suffice. That operator would reconfigure the fluentd pods to tweak the log destination

Reply:
Sounds like an interesting concept, but I am not sure I can fully imagine how will I go about implementing it. That cluster-wide manager daemon will set the `exclude_path` in fluentd or will somewhere in the middle of the pipeline throw away certain logs? What will be the discriminating factor for throwing away unwanted logs? A pod annotation or something else?

Also worth mentioning, one of the requirements was to be able to exclude excessive logs as early as possible so an awry pod does not clog the logging pipeline (prevent DoS of aggregated logging)

Reply:
The exclusion may come on the step where the k8s metadata is already present based on the annotation.
That won't prevent  clogging on the individual nodes (or wherever k8s metadata is attached), but it might prevent problems on elasticsearch ingest.